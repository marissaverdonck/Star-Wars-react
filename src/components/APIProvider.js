import React, { useState } from 'react';
import APIContext from './APIContext';
import useFetchData from './FetchData';

const APIProvider = ({ children }) => {
  const url = 'https://swapi.dev/api/';
  const subjectsObject = useFetchData(url);
  const subjectsArray =
    subjectsObject.data &&
    Object.keys(subjectsObject.data).map(key => ({
      title: key,
      url: url + key
    }));
  const [activeSubject, setActiveSubject] = useState('people');
  const subjectDataObject = useFetchData(url + activeSubject);
  const subjectDataArray = subjectDataObject.data
    ? subjectDataObject.data.results
    : subjectDataObject.data;
  const loadingState = subjectDataObject.loading;
  const errorState = subjectDataObject.error;

  const value = {
    url,
    subjectsArray,
    activeSubject,
    setActiveSubject,
    subjectDataArray,
    loadingState,
    errorState
  };
  return <APIContext.Provider value={value}>{children}</APIContext.Provider>;
};

export default APIProvider;
