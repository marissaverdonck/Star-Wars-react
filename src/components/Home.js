import React, { useContext, lazy, Suspense } from 'react';
import styled from 'styled-components';
import Header from './Header';
import APIContext from './APIContext';
import LoadingState from './LoadingState';
const SubjectData = lazy(() => import('./SubjectData'));

const AppStyle = styled.main`
  background-color: ${({ theme }) => theme.primaryColor};
  min-width: 100vw;
  min-height: 100vh;
  margin: 0;
  padding: 0;
`;

const SelectSubject = styled.section`
  font-family: ${({ theme }) => theme.fontDescription};
  font-size: ${({ theme }) => theme.fontSizeXS};
  margin: 0 auto;
  padding: ${({ theme }) => theme.indentS};
  width: 100vw;
  p {
    font-family: ${({ theme }) => theme.fontDescription};
    font-size: ${({ theme }) => theme.fontSizeXS};
  }
`;
const SelectSubjectElements = styled.div`
  max-width: ${({ theme }) => theme.sectionWidth};
  margin: 0 auto;
  display: flex;
  align-items: center;
`;

const SelectBox = styled.select`
  padding: ${({ theme }) => theme.indentXXS} ${({ theme }) => theme.indentM};
  margin: 0 ${({ theme }) => theme.indentS};
  border: none;
  text-transform: capitalize;
`;

const Home = () => {
  const selectFunction = event => {
    setActiveSubject(event.target.value);
  };
  const { subjectsArray, activeSubject, setActiveSubject } = useContext(
    APIContext
  );

  return (
    <AppStyle>
      <Header size='big' title='Star Wars Universe.' />
      <Suspense fallback={<LoadingState />}>
        <SelectSubject>
          <SelectSubjectElements>
            <p>What are you looking for?</p>
            <SelectBox value={activeSubject} onChange={selectFunction}>
              {subjectsArray &&
                subjectsArray.map((subject, index) => (
                  <option
                    value={subject.title}
                    label={subject.title}
                    key={index}
                  >
                    {subject.title}
                  </option>
                ))}
            </SelectBox>
          </SelectSubjectElements>
        </SelectSubject>
        <SubjectData />
      </Suspense>
    </AppStyle>
  );
};

export default Home;
