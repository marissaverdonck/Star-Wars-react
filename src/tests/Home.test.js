import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Home from '../components/Home';
import APIProvider from '../components/APIProvider';

const valueTypeRight = {
  url: 'https://myurl.com',
  subjectsArray: [1, 2, 3, 4],
  activeSubject: 'fakeactivesubject',
  setActiveSubject: () => {},
  subjectDataArray: [{ name: 'a' }, { title: 'b' }],
  loadingState: true,
  errorState: true
};

const valueTypeWrong = {
  url: 1,
  subjectsArray: { name: 1 },
  activeSubject: 1,
  setActiveSubject: () => {},
  subjectDataArray: { name: 1 },
  loadingState: 1,
  errorState: 1
};

it('Renders correctly when there are right type of values', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <APIProvider value={valueTypeRight}>
          <Home />
        </APIProvider>
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Renders correctly when there are the wrong type of values', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <APIProvider value={valueTypeWrong}>
          <Home />
        </APIProvider>
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
