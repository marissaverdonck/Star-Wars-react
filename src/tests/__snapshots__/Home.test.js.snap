// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly when there are right type of values 1`] = `
<main
  className="sc-hKFyIo jBItgO"
>
  <header
    className="sc-bdnylx boUdXo"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <h1
        className="sc-gtssRu ktzkmZ"
        size="big"
      >
        Star Wars Universe.
         
      </h1>
    </a>
  </header>
  <div
    className="sc-dlnjPT hCuWvg"
  >
    <p>
      Loading data
    </p>
  </div>
</main>
`;

exports[`Renders correctly when there are the wrong type of values 1`] = `
<main
  className="sc-hKFyIo jBItgO"
>
  <header
    className="sc-bdnylx boUdXo"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <h1
        className="sc-gtssRu ktzkmZ"
        size="big"
      >
        Star Wars Universe.
         
      </h1>
    </a>
  </header>
  <section
    className="sc-eCApGN OjKCW"
  >
    <div
      className="sc-jSFkmK ezGZQI"
    >
      <p>
        What are you looking for?
      </p>
      <select
        className="sc-gKAblj evbUBo"
        onChange={[Function]}
        value="people"
      />
    </div>
  </section>
  <section
    className="sc-iCoHVE kSnZbO"
  >
    <div
      className="sc-fujyUd kJyycj"
    />
  </section>
</main>
`;
